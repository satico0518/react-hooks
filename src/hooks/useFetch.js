import { useEffect, useState, useRef } from 'react';

export default (url) => {
  const isMounted = useRef(true);
  useEffect(() => () => (isMounted.current = false), []);
  const initialData = {
    data: [{quote: ''}],
    loading: true,
    error: null,
  };
  const [data, setData] = useState(initialData);

  const getData = async () => {
    setData(initialData);
    try {
      const response = await fetch(url);
      const data = await response.json();
      if (isMounted.current) setData({ data, loading: false, error: null });
    } catch (error) {
      setData({ ...data, error });
    }
  };

  useEffect(() => {
    getData();
  }, [url]);

  return data;
};
