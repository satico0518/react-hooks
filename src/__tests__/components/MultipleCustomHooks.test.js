import React from 'react';

const { shallow } = require('enzyme');
const {
  default: MultipleCustomHooks,
} = require('../../components/03-examples/MultipleCustomHooks');
import '@testing-library/jest-dom';
import useFetch from '../../hooks/useFetch';
jest.mock('../../hooks/useFetch');

describe('MultipleCustomHooks', () => {
  test('should render ok', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
    });
    const wrapper = shallow(<MultipleCustomHooks initialId={5} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('p').text()).toBe('loading...');
  });
  test('should render correct info', () => {
    useFetch.mockReturnValue({
      data: [{ quote: 'Hello world!', author: 'Davo Gomez' }],
      loading: false,
    });
    const wrapper = shallow(<MultipleCustomHooks initialId={5} />);
    expect(wrapper.find('.mb-0').text()).toBe('Hello world!');
    expect(wrapper.find('footer').text()).toBe('Davo Gomez');
  });
  test('should call setId on button click', () => {
    useFetch.mockReturnValue({
      data: [{ quote: 'Hello world!', author: 'Davo Gomez' }],
      loading: false,
    });
    const wrapper = shallow(<MultipleCustomHooks initialId={5} />);
    wrapper.find('button').simulate('click');
    expect(wrapper.find('h2').text()).not.toBe('current Quote Id: 5');
  });
});
