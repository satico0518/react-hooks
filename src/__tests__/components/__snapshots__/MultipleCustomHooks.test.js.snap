// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleCustomHooks should render ok 1`] = `
<Fragment>
  <h2>
    current Quote Id: 
    5
  </h2>
  <button
    onClick={[Function]}
  >
    random Id
  </button>
  <div>
    <p>
      loading...
    </p>
  </div>
</Fragment>
`;
